File: java/selenium/src/test/java/dev/selenium/request/WaitTest.java
package dev.selenium.request;

import dev.selenium.TestBase;

import java.net.URI;
import java.time.Duration;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Predicate;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.remote.http.HttpMethod;
import org.openqa.selenium.remote.http.HttpRequest;
import org.openqa.selenium.support.ui.WebDriverWait;

public class WaitTest extends TestBase {
  @Test
  public void waitAndGetRequest() {
    Predicate<URI> filter = uri -> true;
    AtomicReference<HttpRequest> request = new AtomicReference<>();

    ((FirefoxDriver) driver)
        .network()
        .addRequestHandler(
            filter,
            req -> {
              if (req.getUri().contains("png")) {
                request.set(req);
              }
              return req;
            });

    // .  RequestInterceptOptions options =
    // RequestInterceptOptionsBuilder.addFilter(Map.of("host", "httpbin.org").build();
    //    Handler registration = driver
    //            .network()
    //            .addRequestHandler(options, req -> {
    //                if (req.resourceType() == ResourceType.IMAGE) {
    //                    req.finish()
    //            });

    driver.get("https://selenium.dev");

    WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));

    // It returned something
    wait.until(d -> request.get() != null);
    // It returned the first thing
    Assertions.assertTrue(
        request.get().getUri().contains("sponsors"), "It did not return the first thing");

    // HttpRequest request = wait.until(d -> registration.getResult());
    // Assertions.assertTrue(request.getUri().contains("sponsors"));
  }
}

File: dotnet/Selenium/Request/WaitTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium.BiDi.Network;
using OpenQA.Selenium.Support.UI;

namespace Selenium.Request;

[TestClass]
public class WaitTest : TestBase
{
    [TestMethod]
    public async Task WaitAndGetRequest()
    {
        InterceptedRequest request = null;

        await using var intercept = await Bidi.Network.InterceptRequestAsync(async req =>
        {
            if (req.Request.Url.EndsWith(".png"))
            {
                request = req;
            }

            await req.ContinueAsync();
        });

        // var options = new RequestInterceptOptions {
        //               UrlPatterns = new List<UrlPattern> { new UrlPattern("https://www.selenium.dev/") },
        //           };
        // await using var registration = await driver.Network.AddRequestHandlerAsync(
        //     options,
        //     req =>
        //     {
        //         if (req.Url.EndsWith(".png")
        //         {
        //             req.Complete();
        //         }
        //     });

        await Driver.Navigate().GoToUrlAsync("https://selenium.dev");

        var wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(10));

        // Can return values from the intercept
        wait.Until(x => request != null);
        // request = wait.Until(x => registration.getResult() != null)

        // We want to support the ability to stop processing after the first match
        Assert.IsTrue(request!.Request.Url.Contains("sponsors"), "Processing did not stop on first match");
    }
}